File: src/UnionFind.cpp
//----------------------------------------------------------------------
//                        Disjoint-set data structure 
// File:                        union_find.cpp
//----------------------------------------------------------------------
// Copyright (c) 2018 Matt Piekenbrock. All Rights Reserved.
//
// Class definition based off of data-structure described here:  
// https://en.wikipedia.org/wiki/Disjoint-set_data_structure
#include <Rcpp.h>
using namespace Rcpp; 

#include "UnionFind.h"


SEXP as_XPtr(UnionFind* uf){
  Rcpp::XPtr< UnionFind > p(uf, false);
  return(p);
}
                   
void printCC(UnionFind* uf){
  for (size_t j = 0; j < uf->size; ++j){ Rcout << uf->Find(j) << " "; }
  Rcout << std::endl;
}

// Export as an Rcpp module
RCPP_MODULE(union_find_module) {
  Rcpp::class_<UnionFind>("UnionFind")
  .constructor< std::size_t >()
  .field_readonly( "size", &UnionFind::size)
  .field_readonly( "parent", &UnionFind::parent)
  .field_readonly( "rank", &UnionFind::rank)
  .method("as_XPtr", &as_XPtr)
  .method("print", &printCC )
  .method("connected_components", &UnionFind::ConnectedComponents)
  .method("find", &UnionFind::Find)
  .method("find_all", &UnionFind::FindAll)
  .method("union", &UnionFind::Union)
  .method("union_all", &UnionFind::UnionAll)
  .method("add_sets", &UnionFind::AddSets)
  ;
}

File: R/union_find.R
#' UnionFind
#' @aliases union_find 
#' @description Union find structure exposed as an Rcpp Module.
#' @docType class 
#' @section Methods: 
#' \describe{
#'     \item{$\code{print.simplextree}}{ S3 method to print a basic summary of the simplex tree. }
#' }
#' @author Matt Piekenbrock
#' @import methods
#' @param n Number of elements in the set. 
#' @return A disjoint set, as a \code{Rcpp_UnionFind} object (Rcpp module). 
union_find <- function(n = 0L){
  stopifnot((is.numeric(n) && n >= 0) || is.list(n))
  uf <- new(UnionFind, n)
  return(uf)
}